File: Exercism/cpp/queen-attack/queen_attack.cpp
#include "queen_attack.h"

namespace queen_attack {
	chess_board::chess_board()
	{
		whiteQ = {0, 3};
		blackQ = {7, 3};
	}

	chess_board::chess_board(std::pair<int, int> _w = {0, 3}, std::pair<int, int> _b = {7, 3})
	{
		if(_w == _b)
		{
			throw std::domain_error("no");
		}

		whiteQ = _w;
		blackQ = _b;
	}

	std::pair<int, int> chess_board::white() const
	{
		return whiteQ;
	}

	std::pair<int, int> chess_board::black() const
	{
		return blackQ;
	}

	bool chess_board::can_attack() const
	{
		if(whiteQ.first == blackQ.first || whiteQ.second == blackQ.second)
		{
			return true;
		}
		else if(std::abs(whiteQ.first - blackQ.first) == std::abs(whiteQ.second - blackQ.second))
		{
			return true;
		}
		else
		{
			return false;
		}
	}

	chess_board::operator std::string() const
	{
		std::string b = "";

		for(int i = 0; i < 8; i++)
		{
			for(int j = 0; j < 8; j++)
			{
				if(whiteQ.first == i && whiteQ.second == j)
				{
					b += "W ";
				}
				else if(blackQ.first == i && blackQ.second == j)
				{
					b += "B ";
				}
				else
				{
					j != 7 ? b += "_ " : b += "_";
				}
			}

			b += "\n";
		}

		return b;
	}
}  // namespace queen_attack

File: Exercism/typescript/queen-attack/queen-attack.ts
type Position = readonly [number, number]

type Positions = {
  white: Position
  black: Position
}

export class QueenAttack {
  public readonly black: Position
  public readonly white: Position

  constructor({ white = [7, 3], black = [0, 3] }: Partial<Positions> = {}) {
    this.white = white
    this.black = black

    if(
      (this.white[0] < 0 || this.white[1] < 0 || this.black[0] < 0 || this.black[1] < 0) ||
      (this.white[0] > 7 || this.white[1] > 7 || this.black[0] > 7 || this.black[1] > 7)
    ) {
      throw new Error('Queen must be placed on the board')
    } else if(this.white[0] == this.black[0] && this.white[1] == this.black[1]) {
      throw new Error('Queens cannot share the same space')
    }
  }

  toString(): string {
    let b: string = ''

    for(let i: number = 0; i < 8; i++)
    {
      for(let j: number = 0; j < 8; j++)
      {
        if(this.white[0] == i && this.white[1] == j)
        {
          j != 7 ? b += 'W ' : b += 'W'
        }
        else if(this.black[0] == i && this.black[1] == j)
        {
          j != 7 ? b += 'B ' : b += 'B'
        }
        else
        {
          j != 7 ? b += '_ ' : b += '_';
        }
      }

      if(i < 7) {
        b += '\n'
      }
    }

    return b
  }

  get canAttack() {
    if(this.white[0] == this.black[0] || this.white[1] == this.black[1])
    {
      return true
    }
    else if(Math.abs(this.white[0] - this.black[0]) == Math.abs(this.white[1] - this.black[1]))
    {
      return true
    }
    else
    {
      return false
    }
  }
}
